### PR TITLE
Add blank vote in consultations

### PR DESCRIPTION
**Pull request by LauraJaimeC**
#### :tophat: What? Why?
Add blank vote in consultations and upgrades Decidim from 0.21 to 0.22.

In order to do that:

In admin:
- Add option in Consultations > [Consultation] > Questions > [Question] > Configuration > New checkbox
Captura de pantalla de 2020-12-29 08-42-49

- Add option in Consultations > [Consultation] > Questions > [Question] > Responses > New response > New checkbox
Captura de pantalla de 2020-12-29 08-42-39

In public:
- Add blank vote in votes
Captura de pantalla de 2020-12-29 08-46-02

Captura de pantalla de 2020-12-29 08-44-21

- Add blank votes in results
Captura de pantalla de 2020-12-29 08-46-38

#### :clipboard: Subtasks
- [ ] Add `CHANGELOG` entry
- [ ] Add documentation regarding the feature 
- [ ] Add/modify seeds
- [ ] Add tests
- [ ] Another subtask

### CODE REVIEW

**Review by tramuntanal — COMMENTED** [review]
This PR also upgrades decidim from 0.21 to 0.22, can you add this fact also into de Description please?

**Comment by LauraJaimeC** [comment]
The base branch is changed to master_022.

**Review by ivan-mr — CHANGES_REQUESTED** [review]
In terms of user experience, I have seen that in Spanish, the label "blank vote" always appears in English and not in Spanish "voto en blanco", both in the administrator section and on the voting screen:
Captura de pantalla de 2021-01-11 14-09-10
Captura de pantalla de 2021-01-11 14-10-39

**Comment by ivan-mr** [comment]
Now I have the Spanish language selected on the platform and Catalan appears as the second language in the selector. When pressing on the Catalan language, "blank vote" appears in this:
Captura de pantalla de 2021-01-11 14-10-39

Captura de pantalla de 2021-01-11 14-18-18

**Comment by LauraJaimeC** [comment]
> In terms of user experience, I have seen that in Spanish, the label "blank vote" always appears in English and not in Spanish "voto en blanco", both in the administrator section and on the voting screen:
> 
> Captura de pantalla de 2021-01-11 14-09-10


Captura de pantalla de 2021-01-12 08-34-13

I don't know if this is where you say, but it looks good to me, if I change the language, the translation is changed. If you can tell me exactly what steps you have done.
My steps:
I've created a blank vote response -> I've answered the blank vote -> I've closed the query and published the votes.

**Comment by LauraJaimeC** [comment]
> Now I have the Spanish language selected on the platform and Catalan appears as the second language in the selector. When pressing on the Catalan language, "blank vote" appears in this:
> Captura de pantalla de 2021-01-11 14-10-39
> 
> Captura de pantalla de 2021-01-11 14-18-18

Can you show me how set languages? I have spanish, catalan and english. How can set only two languages?

File: app/assets/javascripts/decidim/admin/blank_vote.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
+const PLACEHOLDER_ES = 'Voto en blanco';
```
Comment (ivan-mr) [inline app/assets/javascripts/decidim/admin/blank_vote.js:1]: Hi @LauraJaimeC ,
I know that Suara now only works with spanish and catalan, but, in the future, if Suara starts to work with another languages different from these three, we will have to remember us to increase this code.
I think there are better approaching than defining the locales in the js.

**Review by ivan-mr — CHANGES_REQUESTED** [review]

File: app/assets/javascripts/decidim/admin/blank_vote.js (hunk 1 above)
Comment (LauraJaimeC) [reply app/assets/javascripts/decidim/admin/blank_vote.js]: Hello @ivan-mr, I've already modified it, can you take a look now?

Thanks!

File: app/overrides/decidim/consultations/admin/responses/add_blank_vote_to_consultations_responses_form.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+Deface::Override.new(virtual_path: 'decidim/consultations/admin/responses/_form',
+                     name: 'add_blank_vote_to_consultations_responses_form',
+                     insert_before: "erb[loud]:contains('form.translated :text_field')",
+                     text: "
+                        <% if current_question.blank_vote %>
+                          <%= form.check_box :blank_vote, { onchange: 'blankVote(this)' }, { data: { 'es': I18n.with_locale(:es) { t('activemodel.attributes.response.blank_vote') },
```
Comment (ivan-mr) [inline app/overrides/decidim/consultations/admin/responses/add_blank_vote_to_consultations_responses_form.rb:8]: I think is better to not consider the available languages like something that will never change. As I see it, it is better to collect the premises directly from the organization and then their corresponding values to assign the blank voting label.

`
<%= form.check_box :blank_vote, { onchange: 'blankVote(this)' }, { data: 
      Hash[*current_question.organization.available_locales.map{|locale| 
         [locale, I18n.with_locale(locale.to_sym) { t('activemodel.attributes.response.blank_vote') }] 
      }.flatten] }.to_json 
%>
`

**Review by ivan-mr — CHANGES_REQUESTED** [review]

File: app/assets/javascripts/decidim/admin/blank_vote.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    getInputTitles(true, translations);
+  } else {
+    getInputTitles(false, translations);
+  }
+}
+
+function getInputTitles (blankVote, translations) {
+  const languageTitles = document.getElementById('response-title-tabs').getElementsByTagName('li');
+
+  for (let i = 0; i < languageTitles.length; i++) {
```
Comment (ivan-mr) [inline app/assets/javascripts/decidim/admin/blank_vote.js:14]: Sorry but I missed check this part of code applying the same criteria.
In this case is better to get the different languages from de data getted from the blank vote checkbox field. With this you can set all locale title fields you need:

`  
for(let i=0; i< Object.keys(translations.data).length; i++) {
    putTitle(Object.values(translations.data)[i], Object.keys(translations.data)[i], blankVote);
  }
`
(I don't know why the code render not properly rendering my code in the comment)

File: app/assets/javascripts/decidim/admin/blank_vote.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+      case 'English':
+        putTitle(translations.data.en, i, blankVote);
+        break;
+    }
+  }
+}
+
+function putTitle(element, index, blankVote) {
+  let title = '';
+  title = document.getElementById(`response-title-tabs-title-panel-${index}`).firstChild;
```
Comment (ivan-mr) [inline app/assets/javascripts/decidim/admin/blank_vote.js:32]: With the approach commented before, we will have to select directly the field that contains the locale identification.

`  
title = document.getElementById('response_title_${locale}');
`

**Review by ivan-mr — CHANGES_REQUESTED** [review]

**Comment by LauraJaimeC** [comment]
Hi @ivan-mr , I change javascript and I fix update response (I forgot it and today I tried it and it didn't work).

File: app/decorators/decidim/consultations/admin/response_form_decorator.rb [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  attribute :question_slug, String
+  attribute :blank_vote
+
+  validate :only_one_blank_vote_response
+
+  def only_one_blank_vote_response
+    question = Decidim::Consultations::Question.find_by(slug: question_slug) if question_slug
+    blank_vote_ = blank_vote == '1' ? true : false
+  
+    if question.responses.where(blank_vote: true).count == 1 && blank_vote_ && id != question.responses.find_by(blank_vote: true).id
```
Comment (ivan-mr) [inline app/decorators/decidim/consultations/admin/response_form_decorator.rb:13]: Also, if you want to make it a little more estable, add on this conditional if question is blank? or also do it like this:
`
if question&.responses&.where(blank_vote: true).count == 1 && blank_vote_ && id != question&.responses&.find_by(blank_vore: true)&.id
`

**Review by ivan-mr — CHANGES_REQUESTED** [review]

File: app/decorators/decidim/consultations/admin/response_form_decorator.rb [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+Decidim::Consultations::Admin::ResponseForm.class_eval do
+  attribute :question_slug, String
+  attribute :blank_vote
+
+  validate :only_one_blank_vote_response
+
+  def only_one_blank_vote_response
+    question = Decidim::Consultations::Question.find_by(slug: question_slug) if question_slug
+    blank_vote_ = blank_vote == '1' ? true : false
```
Comment (ivan-mr) [inline app/decorators/decidim/consultations/admin/response_form_decorator.rb:11]: No need of a ternary conditional:
`
blank_vote = blank_vote == '1'
`

**Review by ivan-mr — CHANGES_REQUESTED** [review]

**Review by ivan-mr — APPROVED** [review]
Great! Thanks @LauraJaimeC

**Pull request merged**